libqtile/backend/wayland/qw/output.c

```diff
@@ -156,3 +157,159 @@ void qw_server_output_new(struct qw_server *server, struct wlr_output *wlr_outpu
+    // so we don't destroy it in this function.
+
+    // Remove previous background
+    qw_output_background_destroy(output);
+
+    // Get dimensions of source image and screen
+    int img_width = cairo_image_surface_get_width(source);
+    int img_height = cairo_image_surface_get_height(source);
+    int o_width, o_height;
+    wlr_output_effective_resolution(output->wlr_output, &o_width, &o_height);
+
+    // Calculate x and y scaling factors
+    double scale_x = (double)o_width / img_width;
+    double scale_y = (double)o_height / img_height;
+
+    // Create a new surface sized to output size
+    cairo_surface_t *wallpaper_surface =
+        cairo_image_surface_create(CAIRO_FORMAT_ARGB32, o_width, o_height);
+    cairo_t *cr = cairo_create(wallpaper_surface);
+
+    switch (mode) {
+
+    // We don't touch the image and draw it at 0, 0
+    case WALLPAPER_MODE_ORIGINAL:
+        break;
+
+    // Fill screen as best as possible while preserving aspect ratio
+    // Image is centered on screen
+    case WALLPAPER_MODE_FILL:
+        cairo_rectangle(cr, 0, 0, o_width, o_height);
+        cairo_clip(cr);
+
+        if ((scale_x * img_height) > o_height) {
+            cairo_translate(cr, 0, -(img_height * scale_x - o_height) / 2);
+            cairo_scale(cr, scale_x, scale_x);
+        } else {
+            cairo_translate(cr, -(img_width * scale_y - o_width) / 2, 0);
+            cairo_scale(cr, scale_y, scale_y);
+        }
+        break;
+
+    // Scale image to screen - don't preserve aspect ratio
+    case WALLPAPER_MODE_STRETCH:
+        cairo_scale(cr, scale_x, scale_y);
+        break;
+
+    // Center image on screen but don't scale image
+    case WALLPAPER_MODE_CENTER:
+        int t_x = (o_width - img_width) / 2;
+        int t_y = (o_height - img_height) / 2;
+        cairo_translate(cr, t_x, t_y);
+        break;
+
+    default:
+        wlr_log(WLR_ERROR, "Unexpected wallpaper mode.");
+        cairo_destroy(cr);
+        cairo_surface_destroy(wallpaper_surface);
+        return;
+    }
+
+    // Paint the modified image to our wallpaper image
+    cairo_set_source_surface(cr, source, 0, 0);
+    cairo_paint(cr);
+
+    // Get data and stride from scaled surface
+    unsigned char *data = cairo_image_surface_get_data(wallpaper_surface);
+    int stride = cairo_image_surface_get_stride(wallpaper_surface);
+
+    // Create wlr_buffer from scaled surface data
+    struct wlr_buffer *buffer = cairo_buffer_create(o_width, o_height, stride, data);
+    if (buffer == NULL) {
+        wlr_log(WLR_ERROR, "Failed to create wlr_buffer from scaled surface");
+        cairo_surface_destroy(wallpaper_surface);
+        return;
+    }
+
+    struct wlr_scene_buffer *scene_buf =
+        wlr_scene_buffer_create(output->server->scene_wallpaper_tree, buffer);
+
+    if (output->background.wallpaper == NULL) {
+        output->background.wallpaper = calloc(1, sizeof(struct qw_output_background_wallpaper));
```

Comment:
Question for coding style:

I tend to prefer the following to avoid binding the type twice, what are you though on this?
```suggestion
        output->background.wallpaper = calloc(1, sizeof *output->background.wallpaper);
```

Reply:
I'll defer to the more experience c coders but, to me, the first one makes it clear what type of object is expected there.